floor/floor/processor/utils.py

```diff
@@ -14,40 +14,63 @@ class clocked(object):
 
         COLORS = [RED, GREEN, BLUE]
 
-        @clocked
+        @clocked(frames_per_beat=1)
         def get_next_frame(self, context):
             self.position += 1
             self.position %= len(COLORS)
             color = COLORS[self.position]
             return [color] * 64
     """
-    def __init__(self, frames_per_beat=1):
+    def __init__(self, frames_per_beat=None, frames_per_second=None):
         """Decorator constructor.
 
         Args
-            frames_per_beat: Period for a single frame. The default value of `1`
+            frames_per_beat: Period for a single frame, based on bpm. For example, a value of `1`
                 will hold `get_next_frame` for the duration of a single beat (quarter
                 note). Any numeric value is allowed; `4` means a whole 4/4 measure, where
                 `0.5` means an eighth note.
+            frames_per_second: Period for a single frame, in clock time. For example, a value
+                of `30` will hold `get_next_frame` for the duration of (1/30.0) seconds, or roughly
+                33ms.
         """
+        if frames_per_beat is None and frames_per_second is None:
+            raise ValueError('Must provide frames_per_beat and/or frames_per_second')
         self.frames_per_beat = frames_per_beat
+        self.frames_per_second = frames_per_second
         self.current_frame = None
-        self.last_time = None
+        self.next_time = None
 
     def __call__(self, fn, *args, **kwargs):
         def new_fn(*args, **kwargs):
             context = args[1]
             now = context.clock
-            bpm = context.bpm or 120.0
-            beats_per_second = 60.0/bpm
-            frame_period = beats_per_second / float(self.frames_per_beat)
 
             # Too soon, return current frame.
-            if self.last_time is not None and (now - self.last_time) < frame_period:
+            if self.current_frame and self.next_time and now < self.next_time:
                 return self.current_frame
 
+            bps_deadline = None
+            if self.frames_per_beat:
+                bpm = context.bpm or 120.0
+                beats_per_second = 60.0/bpm
+                bps_frame_period = beats_per_second / float(self.frames_per_beat)
+                bps_deadline = now + bps_frame_period
+
+            fps_deadline = None
+            if self.frames_per_second:
+                fps_frame_period = 1.0 / self.frames_per_second
```

Comment:
One challenge with that is it becomes effectively static / can no longer dynamically change. I don't think fps-twiddling will be _that_ important to a _normal_ run-of-show, but perhaps very useful when debugging new processors and/or new drivers. Could be micro-optimization territory too (i.e. inarguably better but to ultimately no noticeable effect); maybe we leave it for now and take a closer look in an overall benchmarking pass?

Reply:
Yah, I don't want to micro optimize too much, but I think when its trivial (unlike the non-trivial frames_per_beat comment below) we should take the opportunity.

Since frames_per_second is never directly used or changed, this would be easy to put in the constructor, e.g.:

self.fps_frame_period = 1.0 / frames_per_second

Reply:
Oh! You're right - this is _decorator's_ target FPS, not the show. SGTM.